maro/event_buffer/event_buffer.py

```diff
@@ -202,62 +158,38 @@ def execute(self, tick: int) -> List[Event]:
-            while self._current_index < len(cur_events):
-                event = cur_events[self._current_index]
+            # 1. check if current events match tick.
+            while len(cur_events_list):
+                next_events = cur_events_list.pop()
 
-                if event is None:
+                if next_events is None:
+                    # End of current tick.
                     break
 
-                # 2. check if it is a cascade event and its state,
-                #    we only process cascade events that in pending state
-                if event.category == EventCategory.CASCADE and event.state == EventState.PENDING:
-                    # NOTE: here we return all the cascade events next to current one
-                    result = []
-
-                    for j in range(self._current_index, len(cur_events)):
-                        if cur_events[j].category == EventCategory.CASCADE:
-                            result.append(cur_events[j])
-
-                    return result
+                # Only decision event is a list (even only one item).
+                # NOTE: decision event do not have handlers, and simulator will set its state
+                # to finished after recieved an action.
+                if type(next_events) == list:
+                    return next_events
+                else:
+                    next_events.state = EventState.EXECUTING
 
-                # 3. or it is an atom event, just invoke the handlers
-                if event.state == EventState.FINISHED:
-                    self._current_index += 1
-                    continue
-
-                # 3.1. if handler exist
-                if event.event_type and event.event_type in self._handlers:
-                    handlers = self._handlers[event.event_type]
+                # Invoke handlers.
+                if next_events.event_type and next_events.event_type in self._handlers:
+                    handlers = self._handlers[next_events.event_type]
 
                     for handler in handlers:
-                        handler(event)
-
-                # 3.2. sub events
-                for sub_event in event.immediate_event_list:
-                    if sub_event.event_type in self._handlers:
-                        handlers = self._handlers[sub_event.event_type]
-
-                        for handler in handlers:
-                            handler(sub_event)
-
-                        sub_event.state = EventState.FINISHED
-
-                event.state = EventState.FINISHED
-
-                # remove process event
-                # NOTE: bad performance
-                cur_events[self._current_index] = None
-                self._current_index += 1
+                        handler(next_events)
 
-                self._finished_events.append(event)
+                next_events.state = EventState.FINISHED
 
-            # reset
-            self._current_index = 0
+                if self._event_pool.enabled:
```

Comment:
If the ``event_pool`` is ``enabled``, the ``finished_events`` will always be empty?

Reply:
Yes, if someone need to dump finished event, then event_pool should be disabled.

Reply:
Rename the switch ``self._event_pool.enabled`` to ``disable_finished_events`` and add a description for it.

- disable finished events
- aggressively recycle event instance